c10/cuda/CUDACachingAllocator.cpp

```diff
@@ -64,15 +64,64 @@ constexpr size_t kMinLargeAlloc = 10485760; // allocations between 1 and 10 MiB
 constexpr size_t kRoundLarge = 2097152;     // round up large allocs to 2 MiB
 
 struct DeviceStats {
-  uint64_t   amount_allocated;      // total amount allocated in bytes
-  uint64_t   max_amount_allocated;  // max total amount allocated in bytes
-  uint64_t   amount_cached;         // total amount in cache in bytes
-  uint64_t   max_amount_cached;     // max total amount in cache in bytes
+  uint64_t   total_num_alloc_requests;    // total number of allocation requests received by
```

Comment:
I think the tracked stats may need a bit of rework. Some reasonable questions might be "how many allocated blocks are there now?" "What was the peak number of allocated blocks?". (The first would be answerable by subtracting these two stats, but the second isn't computable from this).

I've been looking at how other allocators track these sorts of things. I'll add some suggestions in a bit.

Reply:
Thanks – happy to incorporate any suggestions.

Reply:
I've been looking at [mimalloc](https://github.com/microsoft/mimalloc/blob/270e765454f98e8bab9d42609b153425f749fff6/include/mimalloc-types.h#L304-L337) and for nearly every stat they track "current", "peak", "allocated", and "freed", which seems useful here too.

Something like:

```
struct Stat {
  int64_t current = 0;
  int64_t peak = 0;
  int64_t allocated = 0;
  int64_t freed = 0;
};

struct DeviceStats {
  Stat allocations;
  Stat allocated_bytes;
  ...
}

void update(Stat& stat, int64_t amount) {
  current += amount;
  peak = max(current, peak);
  if (amount > 0) allocated += amount;
  if (amount < 0) freed += amount;
}

void example() {
  update(stats.allocations, 1); // alternatively stats.allocations.update(1); if you prefer
  update(stats.allocated_bytes, nbytes);
}
```

Stat and DeviceStats should be "public". 

I think the things we might care about are:

allocations: (current/peak/allocated/freed) - how many malloc/free requests have we received?
segments: (current/peak/allocated/freed) - how many underlying allocations have we made to the cuda API.
allocated_bytes: (current/peak/allocated/freed) - how many bytes has the user requested (after rounding block sizes)?
reserved_bytes: (current/peak/allocated/freed) - how many bytes have we reserved from CUDA

Possibly also:

active:  (current/peak/allocated/freed) -- how many blocks are either allocated or have positive event_count (i.e. re-use is delayed due to a stream use)
active_bytes:  (current/peak/allocated/freed) -- total size in bytes of in-use blocks

i.e. roughly what's tracked by "total_num_blocks_released". The condition active >= allocated should always hold, and after an empty cache operation (such as on retry) active == allocated.

split: (current/peak/allocated/freed): number of split blocks

i.e. roughly "total_num_blocks_split" but should also be decreased when blocks are merged. I'm skeptical of the usefulness of this and would lean towards not including it. Most of the time it's completely irrelevant. When you're out-of-memory (and think you shouldn't be), the number of split blocks doesn't tell you much, but the actual heap structure does (see https://gist.github.com/colesbury/0ab4eb58c1d542c5a626040e6714388e)

Simple Counters:

cuda_malloc_retries: how many times has cudaMalloc failed with OOM and we've had to empty the cache and retry. This is `total_num_cache_flushes`. Since it already excludes calls to torch.cuda.empty_cache() (which I think was the correct choice), I think we should make that clear in the name.

In general, I think the choice of names is pretty important here. I'm not entirely confident about these particular choices, but I think it makes sense to try to re-use terminology common to other memory allocators when appropriate. I'd also like to avoid the term "cached" because it's been ambiguous if this has included blocks in-use by the application. (i.e. torch.cuda.memory_cached includes bytes in-use by the application). I think "reserved" is a better term here.